### PR TITLE
Simplify form creation

### PR DESCRIPTION
**Pull request by chrisrichardson**
* Keeps the malloc/free out of user code, by passing a function pointer to the create function.

### CODE REVIEW

File: cpp/dolfin/fem/utils.cpp [hunk 1]
```diff
@@ -512,6 +512,17 @@ fem::get_constants_from_ufc_form(const ufc_form& ufc_form)
   return constants;
 }
 //-----------------------------------------------------------------------------
+std::shared_ptr<fem::Form> fem::create_form(
+    ufc_form* (*fptr)(void),
```
Comment (garth-wells) [inline cpp/dolfin/fem/utils.cpp]: What is `fptr`?

Comment (chrisrichardson) [reply cpp/dolfin/fem/utils.cpp]: `/// @param[in] fptr pointer to a function returning a pointer to ufc_form.`

Comment (garth-wells) [reply cpp/dolfin/fem/utils.cpp]: It's a bit cryptic.

I could also mention that the function it points to takes no arguments.

Comment (chrisrichardson) [reply cpp/dolfin/fem/utils.cpp]: That should be clear from the signature `(void)` - but I admit I don't really like C-style function pointer signatures, as they are... cryptic.

Comment (mrambausek) [reply cpp/dolfin/fem/utils.cpp]: Use std::function<ufc_form*(void)> as argument type and rely on implicit conversion when a C-style function is provided (no difference for caller)?

**Pull request merged**